proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1204 @@
+            qualify(modids, I0)
+        interpretImpItem('module' modids, export_env)
+          = qualify(modids, strip(modids, X))
+
+      This new form of import is allowed only with ``-XLocalModules``.
+
+      For example::
+
+        module A where
+          import qualified B (module X (f, g)) as H
+
+      Suppose ``B`` exports ``{ X.f ↦ (B, X.f), X.g ↦ (Q, g) }``. (Those
+      pairs are original names. This means that ``f`` is defined within
+      a module ``X`` within ``B``, while ``g`` is defined in some other
+      module ``Q`` and only re-exported.) Then this example import statement
+      adds ``{ H.X.f ↦ (B, X.f), H.X.g ↦ (Q, g) }`` to the global
+      environment.
+
+      Note that the import item ``module M (x, y)`` means the same
+      as ``M.x, M.y`` in an import specification.
+      
+   #. An ``import`` statement adds bindings to the global environment,
+      as follows, just as described in **Imports** in the Background.
+      The only difference is that all OEnvs are now QEnvs.
+
+#. **Exports**
+
+   A. The Haskell 2010 Report defines export items this way::
+
+        export ::= qvar
+               |   qtycon [ '(' '..' ')' | '(' cname1 ',' ... ',' cnamen ')' ]
+               |   qtycls [ '(' '..' ')' | '(' qvar1 ',' ... ',' qvarn ')' ]
+               |   'module' modid
+
+
+      With ``-XLocalModules`` specified, expand this definition to be ::
+
+        export ::= qvar [ 'qualified' ]
+               |   qtycon [ '(' '..' ')' | '(' cname1 ',' ... ',' cnamen ')' ] [ 'qualified' ]
+               |   qtycls [ '(' '..' ')' | '(' qvar1 ',' ... ',' qvarn ')' ] [ 'qualified' ]
+               |   'module' [ 'qualified' ] modids [ exports ]
+
+      The nonterminal *exports* (from the `Haskell 2010 Report`_)
+      denotes an export specification: a list of ``export``.
+
+      We must now give these new export items semantics by saying how they
+      are interpreted.    
+
+   #. The interpretation of Haskell98-style export items (that is, without ``qualified``)
+      is unchanged::
+
+        interpretExpItem(haskell98item, ambient_env)
+          = { occ ↦ orig | modids.occ ∈ haskell98item
+                         , modids.occ ↦ orig ∈ ambient_env }
+
+      The only difference here from above is now we write ``modids`` for the qualifier.
+
+      Note that ``occ ↦ orig`` is returned, not the full qualified name. This is
+      unchanged from previously, but it was previously a requirement (in order to
+      be type-correct) to return an ``occ ↦ orig`` binding, while now the type
+      has been generalized to return a QEnv.
+      
+   #. If the export item is a ``qvar``, a ``qtycon``, or a ``qtycls`` (henceforth
+      called ``qual``) followed
+      by ``qualified``, the export
+      QEnv from the current module includes ``qual ↦ orig`` (where ``orig``
+      is the original name found by looking up ``qual`` in the ambient QEnv).
```

Comment:
Please give examples.   Perhaps you intend
```
module M( T qualified ) where
  import X( T )
  import Y( T ) as Wombat
```
```
Does that export X.T or Wombat.T?   (I'm assuming that both X and Y export the *same* T, so referring to plain "T" is unambiguous.)

Or maybe it exports just T, and putting `qualified` is a no-op?   I think that's what you intend, but I found it hard to deduce.

Reply:
It just exports `T`. I've clarified the following paragraph.